include/deal.II/base/vectorization.h

```diff
@@ -26,29 +26,29 @@
 #include <cmath>
 
 // Note:
-// The flag DEAL_II_COMPILER_VECTORIZATION_LEVEL is essentially constructed
+// The flag DEAL_II_VECTORIZATION_WIDTH_IN_BITS is essentially constructed
 // according to the following scheme (on x86-based architectures)
 // #ifdef __AVX512F__
-// #define DEAL_II_COMPILER_VECTORIZATION_LEVEL 3
+// #define DEAL_II_VECTORIZATION_WIDTH_IN_BITS 512
 // #elif defined (__AVX__)
-// #define DEAL_II_COMPILER_VECTORIZATION_LEVEL 2
+// #define DEAL_II_VECTORIZATION_WIDTH_IN_BITS 256
 // #elif defined (__SSE2__)
-// #define DEAL_II_COMPILER_VECTORIZATION_LEVEL 1
+// #define DEAL_II_VECTORIZATION_WIDTH_IN_BITS 128
 // #else
-// #define DEAL_II_COMPILER_VECTORIZATION_LEVEL 0
+// #define DEAL_II_VECTORIZATION_WIDTH_IN_BITS 0
 // #endif
-// In addition to checking the flags __AVX__ and __SSE2__, a CMake test,
-// 'check_01_cpu_features.cmake', ensures that these feature are not only
+// In addition to checking the flags __AVX512F__, __AVX__ and __SSE2__, a CMake
+// test, 'check_01_cpu_features.cmake', ensures that these feature are not only
 // present in the compilation unit but also working properly.
 
-#if DEAL_II_COMPILER_VECTORIZATION_LEVEL > 0
+#if DEAL_II_VECTORIZATION_WIDTH_IN_BITS > 0
 
-#  if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 2 && defined(__SSE2__) && \
+#  if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 256 && defined(__SSE2__) && \
```

Comment:
I don't understand why you need to check for the presence of SSE2 macro here. I know, that was here before...

Reply:
The problem we try to address here is mostly user projects that set the CPU flags on their own. What can happen is that deal.II is compiled with say AVX512, but then a user project forgets to add `-march=native` and falls back to SSE2. This leads to very strange errors as the size of data structures differs between the compiled deal.II code sitting in `libdeal_II.so` and the user code.
The second thing this is supposed to prevent is the case when switching between different CPUs with `-march=native`. One could blame the user for it, but I like to speak out when something goes wrong. I assume we should put this discussion in the file, right?

Reply:
Ah, but you mean specifically the SSE2 macro here, given that x86-64 is the only instruction set right now with 256 bits or more? I think the reasoning for using it was to prevent additional trouble when we have another instruction set (like ARM's SVE) where we could have 256 bits (or 512) but not defined AVX. I guess even that case could be commented - or maybe replaced by a more intuitive `#ifdef __x86_64__` (or the 32 bit equivalent, if anyone cares) to explain what we intend to identify here.

Reply:
My question is why the SSE2 check if necessary. It should also warn if the user compiles deal.II with AVX and the own program without any parallelization (so SSE2 is not defined). I think this should still produce an error, so I would remove ``&& defined(SSE2)`` here.

Reply:
At least on x86-64, `__SSE2__` should always be defined (well, I am not completely sure about MSVC) even if we have explicitly disabled vectorization by e.g. setting `DEAL_II_HAVE_SSE2=OFF` because it is part of x86-64. But I do not insist on this topic, so let me remove it and remember that there is something here - well, the first one to try a non x86 architecture with 256+ bit vectors will get this error :wink: